### PR TITLE
94 Refactor Options

### LINKED ISSUE #94: Refactor options
- [x] Use enum list instead of 12 separate boolean fields
- [x] Refactor init/update to use these enum list

### PR DESCRIPTION
**Pull request by mdingjan**
Will close #94 

The Options class has been refactored a lot. :D

Main changes:
- An option'Map' has been added; the optionMap maps optionNames to booleans. This is not an actual 'Map' as TypeScript doesn't know what a Map is... but I've applied a smart way to get this done. :)
- Because of the Map above, a lot of getters have been removed; there is one general option getter now.
- The update and init function have been merged to one function; init, and the function that was used in the chrome storage has been separated from them.
- Furthermore, some changes in the tests have been applied.

Please review! :)

### CODE REVIEW

File: src/octopeer-github/test/main/Options/OptionsTest.ts [hunk 1]
```diff
@@ -47,51 +48,23 @@ describe("The Options class", function() {
         expect(spyNotify).not.toHaveBeenCalled();
     });
 
-    it("should get the logging value", function() {
-        expect(Options.getLogging()).toBe(true);
+    it("should get the option value", function() {
+        let optionList: string[] = Options.generateOptionList();
+        // HashFile is false by default.
+        let turnedOff = 7;
+        for (let i = 0; i < optionList.length && i !== 7; i++) {
```
Comment (mpsijm) [inline src/octopeer-github/test/main/Options/OptionsTest.ts]: You let the for-loop stop at 7? If `hashFile` is false by default, I think the dnwOptions should also still be tested.

File: src/octopeer-github/main/Options/Options.ts [hunk 2]
```diff
@@ -10,75 +10,48 @@
...
-                    changeObject.doNotWatchOnScreenEvents.newValue : self.doNotWatchOnScreenEvents;
-                self.doNotWatchHoverEvents = changeObject.doNotWatchHoverEvents ?
-                    changeObject.doNotWatchHoverEvents.newValue : self.doNotWatchHoverEvents;
-                self.doNotWatchCommentElements = changeObject.doNotWatchCommentElements ?
-                    changeObject.doNotWatchCommentElements.newValue : self.doNotWatchCommentElements;
-                self.doNotWatchKeyboardShortcutEvents = changeObject.doNotWatchKeyboardShortcutEvents ?
-                    changeObject.doNotWatchKeyboardShortcutEvents.newValue : self.doNotWatchKeyboardShortcutEvents;
-                self.notifyObservers();
+    public syncOptionMap (changeObject: any) {
+        // For every option in optionMap, assign only the new values from the storage to them.
```
Comment (thervh70) [inline src/octopeer-github/main/Options/Options.ts]: As you might know I am no fan of inline comments (especially in such small methods)

File: src/octopeer-github/main/Options/Options.ts [hunk 3]
```diff
@@ -120,102 +92,31 @@ const Options = new (class Options {
...
-    }
-
-    /**
-     * Gets doNotWatch preference about comments, from the chrome storage.
-     * User perspective: Do not watch the comments of my pull request.
-     * @returns {Boolean}
-     */
-    public getDoNotWatchCommentElements() {
-        return this.doNotWatchCommentElements;
+    public generateOptionList(): string[] {
```
Comment (mpsijm) [inline src/octopeer-github/main/Options/Options.ts]: Is/should this method (be) used anywhere outside this class (except for the test)? If not, please consider removing it.

File: src/octopeer-github/main/Options/Options.ts [hunk 4]
```diff
@@ -120,102 +92,31 @@ const Options = new (class Options {
...
+        let options: string[] = [];
+        // For every key in optionMap, push the optionName (thus key) to the to be outputted options array.
```
Comment (thervh70) [inline src/octopeer-github/main/Options/Options.ts]: Same here -> inline comment

File: src/octopeer-github/main/Options/Options.ts [hunk 5]
```diff
@@ -10,75 +10,48 @@
...
-                self.doNotWatchOnScreenEvents = changeObject.doNotWatchOnScreenEvents ?
-                    changeObject.doNotWatchOnScreenEvents.newValue : self.doNotWatchOnScreenEvents;
-                self.doNotWatchHoverEvents = changeObject.doNotWatchHoverEvents ?
-                    changeObject.doNotWatchHoverEvents.newValue : self.doNotWatchHoverEvents;
-                self.doNotWatchCommentElements = changeObject.doNotWatchCommentElements ?
-                    changeObject.doNotWatchCommentElements.newValue : self.doNotWatchCommentElements;
-                self.doNotWatchKeyboardShortcutEvents = changeObject.doNotWatchKeyboardShortcutEvents ?
-                    changeObject.doNotWatchKeyboardShortcutEvents.newValue : self.doNotWatchKeyboardShortcutEvents;
-                self.notifyObservers();
+    public syncOptionMap (changeObject: any) {
```
Comment (mpsijm) [inline src/octopeer-github/main/Options/Options.ts]: Please remove the space before the opening `(` for consistency.

File: src/octopeer-github/main/Options/Options.ts [hunk 6]
```diff
@@ -10,75 +10,48 @@
...
+        let options: string[] = this.generateOptionList();
+        chrome.storage.sync.get(options, this.syncOptionMap);
+        chrome.storage.onChanged.addListener((obj, area) => {if (area === "sync") {this.syncOptionMap(obj); }});
     }
 
     /**
-     * Enables a listener that listens for changes in the sync storage area.
-     * This means that any items that was changed (newValue) is set if changed;
+     * Handles synchronization of the optionMap by retrieving values from the chrome storage on initialization and on update.
+     * @param changeObject set of chrome storage properties.s
```
Comment (mpsijm) [inline src/octopeer-github/main/Options/Options.ts]: There's a typo here :P

**Comment by thervh70** [comment]
Nice change! I really like the new map, however right now there is no autocompletion available. This means that, for every time you want to request an option, you might fail to spell the option correctly. Therefore I would like you to change the way the map works (using enums perhaps)

**Comment by MathiasMeuleman** [comment]
I like it very much, yay for refactoring!
Please consider the comment made by @thervh70 
After that it is ready to be merged

File: src/octopeer-github/test/main/Options/OptionsTest.ts (hunk 1 above)
Comment (mdingjan) [inline src/octopeer-github/test/main/Options/OptionsTest.ts]: Will fix

File: src/octopeer-github/main/Options/Options.ts (hunk 2 above)
Comment (mdingjan) [inline src/octopeer-github/main/Options/Options.ts]: Will fix
I'll add the extra documentation to the general description, if necessary

File: src/octopeer-github/main/Options/Options.ts (hunk 5 above)
Comment (mdingjan) [inline src/octopeer-github/main/Options/Options.ts]: As a space
I want to be deleted
So that I can make @mpsijm happy :D 
Will fix :)

Comment (mpsijm) [inline src/octopeer-github/main/Options/Options.ts]: Yaaay :smile: :tada:

File: src/octopeer-github/main/Options/Options.ts (hunk 6 above)
Comment (mdingjan) [inline src/octopeer-github/main/Options/Options.ts]: Will fix

**Comment by mdingjan** [comment]
Thanks for the feedback! :)
I've resolved it.

File: src/octopeer-github/main/Options/Options.ts (hunk 3 above)
Comment (mdingjan) [inline src/octopeer-github/main/Options/Options.ts]: generateOptionList is used for creating an object containing all chrome storage names.
Just to be sure that this is clear :)

**Comment by mpsijm** [comment]
Thanks for resolving the feedback! :)
I'll merge it.

**Pull request merged**